src/System.Linq.Expressions/src/System/Linq/Expressions/IndexExpression.cs

```diff
@@ -457,51 +456,42 @@ private static void ValidateAccessor(Expression instance, MethodInfo method, Par
 
         private static void ValidateAccessorArgumentTypes(MethodInfo method, ParameterInfo[] indexes, ref ReadOnlyCollection<Expression> arguments)
         {
-            if (indexes.Length > 0)
+            if (indexes.Length != arguments.Count)
             {
-                if (indexes.Length != arguments.Count)
+                throw Error.IncorrectNumberOfMethodCallArguments(method);
+            }
+
```

Comment:
@JonHanna There's also an `else if` clause below that which essentially runs if both collections' counts are 0 (and has the exact same effect), so I don't think it makes a difference.